README.md
# nlhtree_py

## Data Structure

This project is a Python3 implementation of the NLHTree, a data structure
defined
[here](https://jddixon.github.io/xlattice/nlhTree.html)
in some detail.

The NLHTree represents a directory structure as
an indented list.  Directory names appear alone on a line; files
in a given directory are listed at an indent of one space below
the directory name, with the file name followed by the
Secure Hash Algoritm hash of the file contents,
[SHA1](https://en.wikipedia.org/wiki/SHA-1) or
[SHA256](https://en.wikipedia.org/wiki/SHA-2)
as appropriate.  These are printed as 40 or 64 hexadecimal digits
respectively.

What follows is an NLHTree representing the directory `dataDir`.

    dataDir
     data1 34463aa26c4d7214a96e6e42c3a9e8f55727c695
     data2 14193743b265973e5824ca5257eef488094e19e9
     subDir1
      data11 58089ce970b65940dd5bf07703cd81b4306cb8f0
      data12 da39a3ee5e6b4b0d3255bfef95601890afd80709
     subDir2
     subDir3
      data31 487607ec22ee1255cc31c35506c64b1819a48090
     subDir4
      subDir41
       subDir411
        data31 0b57d3ab229a69ce5f7fad62f9fe654fe96c51bb

This has two data files in `dataDir/`, `data1` and `data2`.  Each
of these is followed by its 20-byte/160-bit SHA1 hash in hexadecimal form.
The file names are indented one space more than the directory name.

There are also four subdirectories, `subDir1`, `subDir2`, `subDir3`,
and `subDir4`.  These are at the same indent as the two data files.
Files with the subdirectories and indented one space more.

## Use in BuildLists

The NLHTree is used in the
[BuildList](https://jddixon.github.com/xlattice/buildList.html);
it this context it is prefixed with a title, RSA public key, and
timestamp and then optionally digitally signed.  If signed it can
be used to guarantee the integrity of a file system.  That is, it
can be used to detect any modifications to an existing directory
structure.  Alternatively, it can be used to reconstruct a file
system in systems such as distributed version control systems,
with the content key used to guarantee that the files retrieved
are identical to those in the original form of the directory structure.

As an example, this is the BuildList of a similar directory structure:

    -----BEGIN RSA PUBLIC KEY-----
    MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCzxJ0l1e898G/gBB9zBWUoQ7uw
    8C2Z6OTMJeXrNcTR2ZW7IIMevzYHeR26w+k54Roiv4Oec1uGGom4I7TSxF1QCmfG
    PDaWgvzE4mmwbPCiiYt6cl/y7paG00709ZnbNBjbaaS2Y3gWN+HwiBcENyNrX29i
    P3aQwEB1RNVW8r+SIQIDAQAB
    -----END RSA PUBLIC KEY-----
    sample build list
    2016-09-11 23:13:36
    # BEGIN CONTENT #
    dataDir
     data1 32056bdf38bed17ab7f2bfb37421fa5f4caade71
     data2 80b3b965bdfde312a4eeacc87c42f2a68ad1c7d8
     subDir1
      data11 ca83a024cf4bb9c503f89c86b8819e012d64212d
      data12 da39a3ee5e6b4b0d3255bfef95601890afd80709
     subDir2
     subDir3
      data31 e5ea2d73b3801b38e2add428fa98219c48c69e93
     subDir4
      subDir41
       subDir411
        data31 c4d1f005f36404cf15a00ce00d9a136a35409bc4
    # END CONTENT #

    SuJvEG5zYe5SAnYEHynZXvjIPdY/Fr792ltnwJyPxyg2QO+GCrSfepXnNeIUMJtG5c4zamqsijFZYuAuuhIHCxM1sLcEM5PVNmU/cJT9BLWI952bAqqcB+qaWRcDdSt/tQKZCvzeujZTCa9MsbygN2Wo+ToaIv6dkB21WufyRSs=

The RSA private key corresponding to the public part of the key at the
top of the BuildList is used to generate the digital signature at the
bottom.  Given these two bits of information, the public key and the
digital signature, anyone can verify that the BuildList has not been
tampered with.

## Utilities

### nlh_check_in_data_dir

    usage: nlh_check_in_data_dir [-h] [-b LIST_FILE] [-d DATA_DIR] [-j] [-T] [-V]
                                 [-1] [-2] [-3] [-B] [-u U_PATH] [-v]

    list any files in the NLHTree not present in the directory,

    optional arguments:
      -h, --help            show this help message and exit
      -b LIST_FILE, --list_file LIST_FILE
                            where to write listing (default = list.nlh)
      -d DATA_DIR, --data_dir DATA_DIR
                            path to data directory
      -j, --just_show       show options and exit
      -T, --testing         this is a test run
      -V, --show_version    print the version number and exit
      -1, --using_sha1      using the 160-bit SHA1 hash
      -2, --using_sha2      using the 256-bit SHA2 (SHA256) hash
      -3, --using_sha3      using the 256-bit SHA3 (Keccak-256) hash
      -B, --using_blake2b   using blake2b with a 256-bit digest
      -u U_PATH, --u_path U_PATH
                            path to uDir
      -v, --verbose         be chatty

### nlh_check_in_u_dir

    usage: nlh_check_in_u_dir [-h] [-b LIST_FILE] [-j] [-T] [-V] [-1] [-2] [-3] [-B]
                              [-u U_PATH] [-v]

    given a project directory, write an NLHTree while backing the project up to U

    optional arguments:
      -h, --help            show this help message and exit
      -b LIST_FILE, --list_file LIST_FILE
                            where to write listing (default = list.nlh)
      -j, --just_show       show options and exit
      -T, --testing         this is a test run
      -V, --show_version    print the version number and exit
      -1, --using_sha1      using the 160-bit SHA1 hash
      -2, --using_sha2      using the 256-bit SHA2 (SHA256) hash
      -3, --using_sha3      using the 256-bit SHA3 (Keccak-256) hash
      -B, --using_blake2b   using blake2b with a 256-bit digest
      -u U_PATH, --u_path U_PATH
                            path to uDir
      -v, --verbose         be chatty

### nlh_populate_data_dir

    usage: nlh_populate_data_dir [-h] [-b LIST_FILE] [-j] [-p PATH] [-T] [-V] [-z]
                                 [-1] [-2] [-3] [-B] [-u U_PATH] [-v]

    given an NLHTree and U, recreate the corresponding data directory

    optional arguments:
      -h, --help            show this help message and exit
      -b LIST_FILE, --list_file LIST_FILE
                            where to write listing (default = list.nlh)
      -j, --just_show       show options and exit
      -p PATH, --path PATH  path to data directory
      -T, --testing         this is a test run
      -V, --show_version    print the version number and exit
      -z, --dont_do_it      don't actually do anything, just say what you would do
      -1, --using_sha1      using the 160-bit SHA1 hash
      -2, --using_sha2      using the 256-bit SHA2 (SHA256) hash
      -3, --using_sha3      using the 256-bit SHA3 (Keccak-256) hash
      -B, --using_blake2b   using blake2b with a 256-bit digest
      -u U_PATH, --u_path U_PATH
                            path to uDir
      -v, --verbose         be chatty

### nlh_save_to_u_dir

This is the most frequently used utility.  It scans a data directory,
builds an NLHTree using any of the four supported hashes, and then either
writes the NLHTree out to a file or backs up the input data diretory to
a content-keyed store (U) or both.

    usage: nlh_save_to_u_dir [-h] [-b LISTFILE] [-d DATADIR] [-j] [-T] [-V] [-z]
                             [-1] [-2] [-3] [-B] [-u U_PATH] [-v]

    Given a project directory, write an NLHTree while backing the project up
    to content-keyed store U.

    optional arguments:
      -h, --help            show this help message and exit
      -b LISTFILE, --listFile LISTFILE
                            where to write listing (default = list.nlh)
      -d DATADIR, --dataDir DATADIR
                            path to data directory
      -j, --justShow        show options and exit
      -T, --testing         this is a test run
      -V, --showVersion     print the version number and exit
      -z, --dontDoIt        don't actually do anything, just say what you would do
      -1, --using_sha1      using the 160-bit SHA1 hash
      -2, --using_sha2      using the 256-bit SHA2 (SHA256) hash
      -3, --using_sha3      using the 256-bit SHA3 (Keccak-256) hash
      -B, --using_blake2b   using blake2b with a 256-bit digest
      -u U_PATH, --u_path U_PATH
                            path to uDir
      -v, --verbose         be chatty

## Project Status

A good beta.  All tests succeed.

## On-line Documentation

More information on the **nlhtree_py** project can be found
[here](https://jddixon.github.io/nlhtree_py)


CHANGES
py/nlhtree_py/CHANGES


v0.8.3
    2018-02-27
        * fix setup.py, tox config files                            SLOC 3049
v0.8.2
    2018-02-16
        * mop up missed xlattice.u imports                          SLOC 3054
v0.8.1
    2018-02-13
        * use xlu instead of xlattice.u                             SLOC 3054
v0.8.0
    2018-01-26
        * fixed nlh_save_to_u_dir and example4                      SLOC 3059
v0.7.17
    2018-01-25
        * blake2b support plus sketch of tests                      SLOC 3025
v0.7.16
    2017-12-04
        * pep8, some cleanup, code fixes                            SLOC 2949
v0.7.15
    2017-09-23
        * pep8 etc; add projConfig.toml                             SLOC 2962
v0.7.14
    2017-09-10
        * correct tox.ini, *requirements.txt                        SLOC 2920
v0.7.13
    2017-09-01
        * pep8; add tox.ini, *requirements.txt; minor edits cfg     SLOC 2909
v0.7.12
    2017-07-19
        * rewrite setup.py, MANIFEST.in; edit cfg files             SLOC 2865
v0.7.11
    2017-06-14
        * restructure for pytest                                    SLOC 2877
v0.7.10
    2017-04-26
        * drop dist/* from repos; plus minor edits                  SLOC 2938
v0.7.9
    2017-02-26
        * improved docs                                             SLOC 2961
v0.7.8
    2017-02-10
        * add tarballs to repo                                      SLOC 2959
v0.7.7
    2017-02-08
        * pylinting, some respecification                           SLOC 2634
v0.7.6
    2017-02-05
        * add dump_options; elaborate spec                          SLOC 2603
v0.7.5
    2017-01-18
        * use DirStruc; use PEP 8 rnglib names; pylinting           SLOC 2619
v0.7.4
    2017-01-11
        * adjustments for xlattice_py API changes                   SLOC 2619
v0.7.3
    2016-12-12
        * more pylinting                                            SLOC 2618
v0.7.2
    2016-11-05
        * add utilities to web pages; more PEP 8; minor bug fixes   SLOC 2914
v0.7.1
    2016-10-21
        * rename variables per PEP 8; some minor bug fixes          SLOC 2967
v0.7.0
    2016-10-20
        * rename files and API per PEP 8                            SLOC 2875
v0.6.5
    2016-10-11
        * partial refactor for pylint                               SLOC 2677
v0.6.4
    2016-09-15
        * usingSHA, sha3 support, in all utilities                  SLOC 2645
v0.6.3
    2016-09-12
        XXX INSTALLED BUT NOT COMMITTED
        * remove noisy debug statements
v0.6.2
    2016-09-12
        * fix testIters.py, testWalker.py; sha3 tested              SLOC 2678
v0.6.1
    2016-09-09
        * fix FIX_MEs                                               SLOC 2664
v0.6.0
    2016-09-07
        * correct handling of usingSHA1,2                           SLOC 2545
v0.5.0
    2016-09-01
        * usingSHA1 -> usingSHA (PARTIAL FIX)                       SLOC 2442
v0.4.31
    2016-08-25
        * fix license, setup.py                                     SLOC 2419
v0.4.30
    2016-08-18
        * drop leading and trailing spaces from tree name           SLOC 2394
v0.4.29
    2016-08-15
        * edit down README to refer to xlattice/nlhTree.html        SLOC 2391
v0.4.28
    2016-07-22
        * DIR_FLAT etc now attributes of UDir                       SLOC 2391
v0.4.27
    2016-07-17
        * add NLHTree.dropFromUDir(pathToU) and tests               SLCO 2386
v0.4.26
    2016-07-16
        * skeletal testRmFromU                                      SLOC 2323
v0.4.25
    2016-06-22
        * add ghpDoc/{css,img}/
        * corrected several cases of improper use of defaults       SLOC 2242
v0.4.24
    2016-06-13
        * revised to match xlattice.u API changes                   SLOC 1792
v0.4.23
    2016-05-25
        * add bkp2U, other bash scripts; .dvcz etc                  SLOC 1663
v0.4.22
    2016-05-17
        * changed saveToUDir sig again                              SLOC 1638
v0.4.21
    2016-05-16
        * change saveToUDir signature                               SLOC 1630
v0.4.20
    2016-05-15
        * minor fixes                                               SLOC 1647
        * nlh{CheckIn{DataDir,UDir},PopulateDataDir} fixed          SLOC 1624
v0.4.19
    2016-05-14
        * populateDataDir works, passes tests                       SLOC 1648
v0.4.18
    2016-05-13
        * testCrossFunctions succeeds                               SLOC 1608
v0.4.17
    2016-05-12
        * working NLHTree iterator                                  SLOC 1565
v0.4.16
    2016-05-11
        * add saveToUDir,checkIn{U,Data}Dir,populateDataDir         SLOC 1516
v0.4.15
    2016-05-10
        * copyWalk becomes saveToUDir
        * checkWalk becomes checkInUDir
v0.4.14
    2016-05-10
        * move copyWalk, checkWalk here from BuildList              SLOC 1212
v0.4.13
    2016-05-09
        * incomplete testIters.py succeeds                          SLOC 1158
v0.4.12
    2016-05-08
        * implement walk{File,String,Strings} and tests             SLOC 1030
v0.4.11
    2016-05-07
        * add testWalker                                            SLOC  892
v0.4.10
    2016-05-02
        * add clone() for tree and leaf                             SLOC  851
v0.4.9
    2016-04-28
        * corrected countem, projCfg{,.local}                       SLOC  824
v0.4.8
    2016-04-20
        * correct use of exRE; fix countem                          SLOC  801
v0.4.7
    2016-04-13
        * correct project name in setup.py; add .fileMap{,.local}   SLOC  801
v0.4.6
    2016-03-31
        * minor cosmetic edits                                      SLOC  777
v0.4.5
    2016-03-23
        * add auto8it and execute                                   SLOC  777
v0.4.4
    2016-02-25
        * update LICENSE                                            SLOC  757
v0.4.3
    2016-02-01
        * minor edits                                               SLOC  757
v0.4.2
    2015-12-29

testNLHTree3.py
#!/usr/bin/env python3

# testNLHTree3.py

import hashlib
import re
import time
import unittest

from rnglib import SimpleRNG
from nlhtree import NLHTree as NT
from xlattice import (Q, checkUsingSHA,
                      SHA1_BIN_LEN, SHA2_BIN_LEN, SHA3_BIN_LEN,
                      SHA1_HEX_LEN, SHA2_HEX_LEN, SHA3_HEX_LEN)
from xlattice.crypto import SP


class TestNLHTree3 (unittest.TestCase):

    # adapted from the buildList example 2015-05-22
    EXAMPLE1 = [
        'dataDir',
        ' data1 bea7383743859a81b84cec8fde2ccd1f3e2ff688',
        ' data2 895c210f5203c48c1e3a574a2d5eba043c0ec72d',
        ' subDir1',
        '  data11 cb0ece05cbb91501d3dd78afaf362e63816f6757',
        '  data12 da39a3ee5e6b4b0d3255bfef95601890afd80709',
        ' subDir2',
        ' subDir3',
        '  data31 8cddeb23f9de9da4547a0d5adcecc7e26cb098c0',
        ' subDir4',
        '  subDir41',
        '   subDir411',
        '    data41 31c16def9fc4a4b6415b0b133e156a919cf41cc8',
        ' zData 31c16def9fc4a4b6415b0b133e156a919cf41cc8',
    ]
    # this is just a hack but ...
    EXAMPLE2 = [
        'dataDir',
        ' data1 012345678901234567890123bea7383743859a81b84cec8fde2ccd1f3e2ff688',
        ' data2 012345678901234567890123895c210f5203c48c1e3a574a2d5eba043c0ec72d',
        ' subDir1',
        '  data11 012345678901234567890123cb0ece05cbb91501d3dd78afaf362e63816f6757',
        '  data12 012345678901234567890123da39a3ee5e6b4b0d3255bfef95601890afd80709',
        ' subDir2',
        ' subDir3',
        '  data31 0123456789012345678901238cddeb23f9de9da4547a0d5adcecc7e26cb098c0',
        ' subDir4',
        '  subDir41',
        '   subDir411',
        '    data41 01234567890123456789012331c16def9fc4a4b6415b0b133e156a919cf41cc8',
        ' zData 01234567890123456789012331c16def9fc4a4b6415b0b133e156a919cf41cc8',
    ]
    EXAMPLE3 = [
        'dataDir',
        ' data1 6d57759cf499a8ff7762a10043548f22513ed83456452332a8abd4b59d7e9203',
        ' data2 dacbf5c11f4ddbd1277ecbc304e09967d3124148560f82634d3912db8b4bd547',
        ' subDir1',
        '  data11 fb47958129f261f65c1655002ff5f9806bc969283ad772af5e8caaf214a9ed72',
        '  data12 e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855',
        ' subDir2',
        ' subDir3',
        '  data31 7659fb836a76fb3f3369e1a4ca247104220e4778d5862e38a123e10f02520e87',
        ' subDir4',
        '  subDir41',
        '   subDir411',
        '    data41 00bb6d0864cb4952a0c41cbea65cf09de41e00fc6fa1011a27c5dd8814c98175',
    ]

    def setUp(self):
        self.rng = SimpleRNG(time.time())

    def tearDown(self):
        pass

    def doTestPatternMatching(self, usingSHA):
        checkUsingSHA(usingSHA)
        if usingSHA == Q.USING_SHA1:
            ss = self.EXAMPLE1
        elif usingSHA == Q.USING_SHA2:
            ss = self.EXAMPLE2
        elif usingSHA == Q.USING_SHA3:
            ss = self.EXAMPLE3

        # first line --------------------------------------
        m = NT.DIR_LINE_RE.match(ss[0])
        self.assertTrue(m)
        self.assertEqual(len(m.group(1)), 0)
        self.assertEqual(m.group(2), 'dataDir')

        # simpler approach ----------------------
        name = NT.parseFirstLine(ss[0])
        self.assertEqual(name, 'dataDir')

        # file with indent of 1 ---------------------------
        if usingSHA == Q.USING_SHA1:
            m = NT.FILE_LINE_RE_1.match(ss[1])
        else:
            # XXX This works for both SHA2 and SHA3
            m = NT.FILE_LINE_RE_2.match(ss[1])
        self.assertTrue(m)
        self.assertEqual(len(m.group(1)), 1)
        self.assertEqual(m.group(2), 'data1')

        # that simpler approach -----------------
        indent, name, hash = NT.parseOtherLine(ss[1])
        self.assertEqual(indent, 1)
        self.assertEqual(name, 'data1')
        if usingSHA == Q.USING_SHA1:
            self.assertEqual(len(hash), SHA1_HEX_LEN)
        else:
            # XXX This works for both SHA2 and SHA 3
            self.assertEqual(len(hash), SHA2_HEX_LEN)

        # subdirectory ------------------------------------
        m = NT.DIR_LINE_RE.match(ss[3])
        self.assertTrue(m)
        self.assertEqual(len(m.group(1)), 1)
        self.assertEqual(m.group(2), 'subDir1')

        # that simpler approach -----------------
        indent, name, hash = NT.parseOtherLine(ss[3])
        self.assertEqual(indent, 1)
        self.assertEqual(name, 'subDir1')
        self.assertEqual(hash, None)

        # lower level file ----------------------
        if usingSHA == Q.USING_SHA1:
            m = NT.FILE_LINE_RE_1.match(ss[12])
        else:
            # XXX This works for both SHA2 and SHA 3
            m = NT.FILE_LINE_RE_2.match(ss[12])
        self.assertTrue(m)
        self.assertEqual(len(m.group(1)), 4)
        self.assertEqual(m.group(2), 'data41')

        # that simpler approach -----------------
        indent, name, hash = NT.parseOtherLine(ss[12])
        self.assertEqual(indent, 4)
        self.assertEqual(name, 'data41')
        if usingSHA == Q.USING_SHA1:
            self.assertEqual(len(hash), SHA1_HEX_LEN)
        else:
            # XXX This works for both SHA2 and SHA 3
            self.assertEqual(len(hash), SHA2_HEX_LEN)

    def testPatternMatching(self):
        for using in [Q.USING_SHA1, Q.USING_SHA2, Q.USING_SHA3, ]:
            self.doTestPatternMatching(using)

    def doTestSerialization(self, usingSHA):
        checkUsingSHA(usingSHA)
        tree = NT.createFromStringArray(self.EXAMPLE1, usingSHA)
        self.assertEqual(tree.usingSHA, usingSHA)

        ss = []
        tree.toStrings(ss, 0)

        tree2 = NT.createFromStringArray(ss, usingSHA)
        self.assertEqual(tree, tree2)

        s = '\n'.join(ss) + '\n'
        tree3 = NT.parse(s, usingSHA)
        s3 = tree3.__str__()

        self.assertEqual(s3, s)
        self.assertEqual(tree3, tree)

        dupe3 = tree3.clone()
        self.assertEqual(dupe3, tree3)

    def testSerialization(self):
        for using in [Q.USING_SHA1, Q.USING_SHA2, Q.USING_SHA3, ]:
            self.doTestSerialization(using)


if __name__ == '__main__':
    unittest.main()
